enhancements/olm-deprecate-copied-csvs.md

```diff
@@ -0,0 +1,207 @@
+
+### User Stories
+
+#### Toggle Copied CSVs Off
+
+As a Cluster Admin, I want a way to disable Copied CSVs so that I can reduce resource consumption ahead of the feature's deprecation and eventual removal.
+
+#### Toggle Copied CSVs On
+
+As a Cluster Admin, I want a way to enable Copied CSVs in the event users of my cluster are still dependent on them.
+
+#### View "Available" Operators
+
+As a User, I want a way to view the operators that can act in a given namespace I have access to.
+
+### Implementation Details/Notes/Constraints
+
+#### Operator Cardinality
+
+Today, Operator resources are generated on a per-Subscription basis. Since CSVs -- and thus Copied CSVs -- can exist without matching Subscriptions, OLM's controllers will be extended to generate additional Operator resources for these "free CSVs" such that:
+
+- there exists at least one`Operator` resource for every CSV
+- there exists at least one `Operator` resource for every `Subscription`
+
+That is, all CSVs installed by a given `Subscription` share an `Operator` resource __and__ any CSVs that weren't installed by a `Subscription` are related to their own `Operator` resource.
+
+#### Orphaned CSVs
+
+Orphaned CSVs are CSVs that were installed via a `Subscription` that no longer exists. Since the lifetimes -- i.e. the period of existence on a cluster --  of `Operator` resources aren't directly tied to the lifetimes of the resources that triggered their generation, the `Operator` resource(s) associated with Orphaned CSVs will continue to exist _after_ a `Subscription` is deleted. If that `Operator` resource is deleted explicitly, OLM will generate another as per the rules [outlined above](#operator-cardinality).
+
+#### Copied RBAC 
+
+In addition to operator discoverability, Copied CSVs also serve as convienient vehicles for the garbage collection of Copied RBAC. These are resources that are copied along with the CSV to provide the operator access to a namespace. To support the loss of Copied CSVs, the RBAC copying logic in OLM will be extended to write additional owner references from copied resources to the related Operator resource. 
+
+#### The Toggle
+
+A feature toggle will be added to control OLM's use of Copied CSVs. When enabled (toggled on), OLM will delete all existing Copied CSVs and will not generate any more. When disabled (toggled off), OLM will generate Copied CSVs (much like it does today).
+
+When toggled (on then off), OLM will recreate any missing Copied CSVs and reassociate the owner references of related copied resources with them (i.e. the inverse operation of the one described in the [Copied RBAC](#copied-rbac) section).
+
+The toggle will be disabled by default.
+
+The toggle can be set from either a command-line option on the olm-operator binary _or_ a field of the `olm-config` `ConfigMap` colocated in the same namespace as OLM's deployment on-cluster; when both are present, the field of the `ConfigMap` takes precedence.
```

Comment:
- It's a net-new `ConfigMap` to carry config for OLM
- I don't want to include a standalone CRD because: 1. I think it will be more palatable to backport; 2. singleton k8s types feel kludgy

Reply:
A general configuration mechanism has applicability in other areas as well. For example @estroz has been looking for a mechanism to feature-gate new resolver and constraint features.

There are definitely pros and cons of configmap vs CRD, which mostly boil down to configmaps are schemaless, CRDs are (generally) not.

This topic on its own seems deserving on an enhancement, IMO.

Reply:
On point 2. I don't think I'd call this a singleton CRD. I'd expect that the OLM deployment would have a flag that tells OLM the name of the custom object it should read (and watch?). Then if there are somehow multiple instances of OLM installed on the cluster, each of them could point to a different config object.

Reply:
> singleton k8s types feel kludgy

kludgy or not, they are the pattern we've implemented for many other CVO operators at this point, and i haven't seen any issues/downsides/confusion caused by them (e.g. we don't have users filing bug reports because they created a 2nd instance w/ a different name and it was ignored).  That mostly is because:
1) we document the mechanism well (the resource type and specific name to use)
2) we ensure it is pre-created during install so the user only ever has to edit it, not create it (thus they don't have to get the name right)

Reply:
These are fair points. I think you've more or less convinced me that it's reasonable to add a net new API for configuration in OLM for a new new release, but I'm still wondering whether that's a backport-palatable design.

Are we at an impasse where adding a CRD is acceptable for backport? Are there any alternatives I've not considered?

Reply:
do we think we need to backport this?  i know it's a pain point, but is it such a pain point that we can't ask people who care about the problem to upgrade?

while adding a CRD definitely feels like a bright line we wouldn't want to cross on backporting, even if we don't add a CRD this is fundamentally a new feature we're talking about adding and would need to think long and hard about backporting (especially when you factor in potential console impact)

Reply:
> while adding a CRD definitely feels like a bright line we wouldn't want to cross on backporting

i'm actually going to walk this back.  i'm not sure that "it introduces a new CRD/api" is the most important consideration when debating if we'd make an exception to backport a feature.

Reply:
Alright!

So I've updated the EP and swapped the ConfigMap for a newly introduced config CRD and singleton.

PTAL.